### PR TITLE
Better JSON number parsing.

### PR DESCRIPTION
**Pull request by nicolasnoble**
Added some tweaks in the json number parsing to better support valid and invalid cases.

### CODE REVIEW

File: src/core/lib/json/json_reader.c [hunk 1]
```diff
@@ -100,10 +100,35 @@ void grpc_json_reader_init(grpc_json_reader *reader,
...
+  }
+  return success;
+}
+
 int grpc_json_reader_is_complete(grpc_json_reader *reader) {
-  return ((reader->depth == 0) &&
+  return (((reader->depth == 0) &&
           ((reader->state == GRPC_JSON_STATE_END) ||
-           (reader->state == GRPC_JSON_STATE_VALUE_END)));
+           (reader->state == GRPC_JSON_STATE_VALUE_END))) ||
```
Comment (ctiller) [inline src/core/lib/json/json_reader.c]: Suggest breaking this up into more than one line.

Comment (ctiller) [inline src/core/lib/json/json_reader.c]: Maybe with a switch or something.

File: src/core/lib/json/json_reader.c [hunk 2]
```diff
@@ -100,10 +100,40 @@ void grpc_json_reader_init(grpc_json_reader *reader,
...
+}
+
 int grpc_json_reader_is_complete(grpc_json_reader *reader) {
-  return ((reader->depth == 0) &&
-          ((reader->state == GRPC_JSON_STATE_END) ||
-           (reader->state == GRPC_JSON_STATE_VALUE_END)));
+  switch (reader->state) {
+    case GRPC_JSON_STATE_END:
+    case GRPC_JSON_STATE_VALUE_END:
+      return reader->depth == 0;
```
Comment (jboeuf) [inline src/core/lib/json/json_reader.c]: Please comment: 
`// Fall-through intended.`

File: test/core/json/json_test.c [hunk 3]
```diff
@@ -139,6 +142,13 @@ static testing_pair testing_pairs[] = {
...
+    {"[-]", NULL},
+    {"[-01]", NULL},
+    {"[0.3e+]", NULL},
+    {"[1.0e+]", NULL},
+    {"[1.0e-]", NULL},
+    {"[-012]", NULL},
+    {"[-.123]", NULL},
     {"1.", NULL},
     {"1e", NULL},
     {".12", NULL},
```
Comment (jboeuf) [inline test/core/json/json_test.c]: Maybe add `012` as a test case as well.

**Comment by jtattermusch** [comment]
@nicolasnoble  is this PR still useful? Last update is from 2016. Let's resolve/close.

**Comment by jtattermusch** [comment]
Feel free to reopen.

**Pull request closed without merging**